project/src/com/google/daggerquery/executor/models/Query.java

```diff
@@ -260,4 +265,60 @@ private boolean findSomePath(String source, String target, Path path,
 
     return false;
   }
+
+  /**
+   * Checks if the passed {@code node} is in the {@link BindingGraph} or if the user misspelled the node's name.
+   *
+   * <p>If the passed node is correct, does nothing.
+   * Otherwise, it throws an exception, the type of which depends on the node's name.
+   *
+   * @throws MisspelledNodeNameException if specified source node contains a typo
+   * and can be corrected in no more than {@code MAX_NUMBER_OF_MISPLACED_LETTERS} steps
+   *
+   * @throws IllegalArgumentException if specified source node doesn't exist
+   */
+  private void checkNodeForCorrectness(String node, BindingGraph bindingGraph) {
+    if (bindingGraph.getAdjacencyListMap().containsKey(node)) {
+      return;
+    }
+
+    // The specified node could not be found on the graph, we need to check for typos.
+    List<String> closestNodes = findNodesWithClosestName(node, bindingGraph.getAdjacencyListMap().keySet());
+    if (closestNodes.isEmpty()) {
+      throw new IllegalArgumentException(String.format("Specified source node %s doesn't exist.", node));
+    } else {
+      throw new MisspelledNodeNameException(/*nodeNameWithTypo =*/ node, /*correctNodeName =*/ closestNodes);
+    }
+  }
+
+  /**
+   * Finds the closest nodes in a given {@link Set<String>} to the {@code originalNode}.
+   *
+   * <p>For measuring a distance uses <a href = "https://en.wikipedia.org/wiki/Levenshtein_distance">Levenshtein distance metric</a>.
+   * It simply calculates the number of changes required to be made to get one sequence from another,
+   * where each change is a single character modification (deletion, insertion or substitution).
+   *
+   * <p>A node can be considered a neighbor only if the distance between it and the given {@code originalNode}
+   * is less than or equal to {@code MAX_NUMBER_OF_MISPLACED_LETTERS}.
+   *
+   * <p>If no nodes with a similar name are found, returns an empty {@link List<String>}.
+   * Otherwise, it returns a {@link List<String>} with nodes with the same distance.
+   */
+  private List<String> findNodesWithClosestName(String originalNode, Set<String> allNodes) {
+    List<String> closestNodes = new ArrayList<>();
+    int bestDistance = Integer.MAX_VALUE;
+
+    for (String node: allNodes) {
+      int distance = levenshteinDistanceCalculator.apply(originalNode, node);
+      if (distance <= MAX_NUMBER_OF_MISPLACED_LETTERS && distance < bestDistance) {
```

Comment:
Now that we're looking for all of the nodes within 3 steps of the user input (vs the closest node) do we still need this best distance check?

Reply:
Actually no, let's save all suggestions with the distance that works for us 👌